### PR TITLE
sink/mq(cdc): Fix mq flush worker deadlock

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: ref #4987

### What is changed and how it works?
- Make sure that `mqFlushWorker` does not block infinitely if Sarama throws us an error.

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
None.
```

### CODE REVIEW

File: cdc/sink/mq_flush_worker.go [hunk 1]
```diff
@@ -177,7 +186,17 @@ func (w *flushWorker) asyncSend(
 
 // run starts a loop that keeps collecting, sorting and sending messages
 // until it encounters an error or is interrupted.
-func (w *flushWorker) run(ctx context.Context) error {
+func (w *flushWorker) run(ctx context.Context) (retErr error) {
+	defer func() {
+		select {
+		case w.errCh <- retErr:
+		default:
+			log.L().Warn("flushWorker error is dropped due to a previous erro", zap.Error(retErr))
```
Comment (3AceShowHand) [inline cdc/sink/mq_flush_worker.go:194]: how about follow the log calling convention, just call `Warn` directly ?

Comment (liuzix) [reply cdc/sink/mq_flush_worker.go:194]: my mistake

File: cdc/sink/mq_flush_worker.go [hunk 2]
```diff
@@ -17,6 +17,8 @@ import (
 	"context"
 	"time"
 
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+
```
Comment (hi-rustin) [inline cdc/sink/mq_flush_worker.go:21]:
```suggestion
```

File: cdc/sink/mq_flush_worker.go [hunk 3]
```diff
@@ -61,6 +69,7 @@ type flushWorker struct {
 func newFlushWorker(encoder codec.EventBatchEncoder, producer producer.Producer, statistics *Statistics) *flushWorker {
 	w := &flushWorker{
 		msgChan:    make(chan mqEvent),
+		errCh:      make(chan error, 1), // must be buffered chan
```
Comment (hi-rustin) [inline cdc/sink/mq_flush_worker.go:72]: Can you explain it in more detail?

File: cdc/sink/mq_flush_worker_test.go [hunk 4]
```diff
@@ -413,3 +429,45 @@ func TestAbort(t *testing.T) {
 	cancel()
 	wg.Wait()
 }
+
+func TestProducerError(t *testing.T) {
+	t.Parallel()
+
+	worker, _ := newTestWorker()
```
Comment (hi-rustin) [inline cdc/sink/mq_flush_worker_test.go:436]: You can get the producer here.

File: errors.toml [hunk 5]
```diff
@@ -501,6 +501,11 @@ error = '''
 unknown '%s' protocol for Message Queue sink
 '''
 
+["CDC:ErrMQWorkerClosed"]
+error = '''
+MQ worker has closed
```
Comment (hi-rustin) [inline errors.toml:506]:
```suggestion
MQ flush worker has closed
```

**Review by 3AceShowHand — APPROVED** [review]

**Review by hi-rustin — APPROVED** [review]

File: cdc/sink/mq_flush_worker.go [hunk 6]
```diff
@@ -177,7 +187,17 @@ func (w *flushWorker) asyncSend(
...
 // until it encounters an error or is interrupted.
-func (w *flushWorker) run(ctx context.Context) error {
+func (w *flushWorker) run(ctx context.Context) (retErr error) {
+	defer func() {
+		select {
+		case w.errCh <- retErr:
+		default:
+			log.Warn("flushWorker error is dropped due to a previous error", zap.Error(retErr))
+		}
+		close(w.errCh)
```
Comment (amyangfei) [inline cdc/sink/mq_flush_worker.go:197]: Have a question here, could the `default` branch run?
Supposing the err channel is full in L193, the channel could have been closed, the panic `send on closed channel` could happen.

Comment (liuzix) [reply cdc/sink/mq_flush_worker.go]: The channel is written to exactly once.

**Review by amyangfei — APPROVED** [review]

**Comment by hi-rustin** [comment]
/hold 

The `multi_topics` test seems to be broken.

/run-kafka-integration-test

**Comment by hi-rustin** [comment]
> /hold
> 
> The `multi_topics` test seems to be broken.
> 
> /run-kafka-integration-test

It should be flaky test, I added some retry times. Because there are three different consumers writing at the same time.

**Pull request merged**